### PR TITLE
Feature/photon os bootstrap

### PR DESCRIPTION
**Pull request by cgx027**
Photon OS bootstrap support

This is one of three PRs for Photon OS bootstrap:
RackHD/on-tasks #275  https://github.com/RackHD/on-tasks/pull/275
RackHD/on-taskgraph #138 https://github.com/RackHD/on-taskgraph/pull/138 
RackHD/on-http #362 https://github.com/RackHD/on-http/pull/362

@RackHD/corecommitters @iceiilin @pengz1 @WangWinson

### CODE REVIEW

File: data/templates/photon-os-ks [hunk 1]
```diff
@@ -0,0 +1,35 @@
+{
+    "hostname": "test_hostname",
```
Comment (yyscamper) [inline data/templates/photon-os-ks]: Use template hostname: <%=hostname%> and the `hostname` may be optional, see https://github.com/RackHD/on-http/blob/master/data/templates/centos-ks#L127

File: data/templates/photon-os-ks [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    "password":
+        {
+            "crypted": false,
```
Comment (yyscamper) [inline data/templates/photon-os-ks]: Why not use the crypted password?

File: data/templates/photon-os-ks [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+                    "chmod 0644 $filename",
+
+                    "# Install the service",
+                    "systemctl enable <%=rackhdCallbackScript%>.service",
+
+                    "# Enable SSH root login",
+                    "sed -i 's/PermitRootLogin no/PermitRootLogin yes/g' /etc/ssh/sshd_config",
+
+                    "# Signify ORA the installation completed",
+                    "curl http://<%=server%>:<%=port%>/api/common/templates/renasar-ansible.pub"
```
Comment (yyscamper) [inline data/templates/photon-os-ks]: the renasar-ansible.pub should be <%=completionUri%>

File: data/templates/photon-os.rackhdcallback [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+# Short-Description: Callback to rackhd post installation API hook
+# Description: Callback to rackhd post installation API hook
+### END INIT INFO
+
+
+# We can't really know when networking is actually up and running from
+# a simple "Required-Start: $network", so instead just use curl with
+# a bunch of retries and hope it works. We could use more sophisticated
+# dependency mechanisms provided by systemd, but ideally we can just use
+# a single script to service both CentOS 6.5 and CentOS 7 installs, hence
```
Comment (yyscamper) [inline data/templates/photon-os.rackhdcallback]: The comment is CentOS, but you are working on Photon OS

File: data/templates/photon-os.rackhdcallback [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+# reasoning for using `wget --retry-connrefused` below. Using `curl --retry`
```
Comment (yyscamper) [inline data/templates/photon-os.rackhdcallback]: You are using `curl`, but the comment is wget

File: data/templates/photon-os.rackhdcallback [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+# fails immediately because it only handles connection timeouts, not connection
+# refused cases.
+# See https://www.freedesktop.org/wiki/Software/systemd/NetworkTarget/
+
+# Nothing wrong with set -e here since we're not doing anything complex
+# set -e
```
Comment (yyscamper) [inline data/templates/photon-os.rackhdcallback]: line 26 and 27 should be removed.

File: data/templates/photon-os.rackhdcallback [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+        /bin/curl --max-time 1000 --connect-timeout 9 --retry-delay 1 --retry 100 http://<%=server%>:<%=port%>/api/common/templates/<%=completionUri%>
+
+        # Only run this once to verify the OS was installed, then stop the service and remove itself forever
+        systemctl disable photon-os.rackhdcallback.service
+        rm /etc/photon-os.rackhdcallback
+        rm /etc/systemd/system/photon-os.rackhdcallback.service
+        exit
+	fi
+	
+    sleep $interval
```
Comment (yyscamper) [inline data/templates/photon-os.rackhdcallback]: indent error

File: data/templates/photon-os-ks (hunk 1 above)
Comment (cgx027) [inline data/templates/photon-os-ks]: Thanks @yyscamper , changed to:

```<% if (typeof hostname !== 'undefined') { %>"hostname": "<%=hostname%>",<% } %>```

File: data/templates/photon-os-ks (hunk 3 above)
Comment (cgx027) [inline data/templates/photon-os-ks]: Good finding, thanks. 

Changed to ```"curl http://<%=server%>:<%=port%>/api/common/templates/<%=completionUri%>" ```

File: data/templates/photon-os-ks (hunk 2 above)
Comment (cgx027) [inline data/templates/photon-os-ks]: when set crypted to ture, the following password text will be crypted, which make it hard for user to set a password.

File: data/templates/photon-os-ks [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+                    "# Install the service",
+                    "systemctl enable <%=rackhdCallbackScript%>.service",
+
+                    "# Enable SSH root login",
+                    "sed -i 's/PermitRootLogin no/PermitRootLogin yes/g' /etc/ssh/sshd_config",
+
+                    "# Signify ORA the installation completed",
+                    "curl http://<%=server%>:<%=port%>/api/common/templates/<%=completionUri%>"
+                   ]<% if (typeof rootSshKey !== 'undefined') { %>,
+    "public_key": <%=rootSshKey%><% } %>
```
Comment (iceiilin) [inline data/templates/photon-os-ks]: installPartition hasn't been added here. I will modify later.

Comment (iceiilin) [inline data/templates/photon-os-ks]: What if rootSshKey isn't specified in options?

File: data/templates/photon-os-ks [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+
+                    "# Install the service",
+                    "systemctl enable <%=rackhdCallbackScript%>.service",
+
+                    "# Enable SSH root login",
+                    "sed -i 's/PermitRootLogin no/PermitRootLogin yes/g' /etc/ssh/sshd_config",
+
+                    "# Signify ORA the installation completed",
+                    "curl http://<%=server%>:<%=port%>/api/common/templates/<%=completionUri%>"
+                   ]<% if (typeof rootSshKey !== 'undefined') { %>,
```
Comment (iceiilin) [inline data/templates/photon-os-ks]: Would it be more clear if moving the if statement to a new line?

**Comment by WangWinson** [comment]
Since we are now support read sub folder of `/template`, suggest create a `/template/install-photon` folder and put related template in.

File: data/templates/photon-os-ks [hunk 10]
```diff
@@ -0,0 +1,35 @@
+{
+    <% if (typeof hostname !== 'undefined') { %>"hostname": "<%=hostname%>",<% } %>
```
Comment (WangWinson) [inline data/templates/photon-os-ks]: More readable as below:
```javascript
<% if (typeof hostname !== 'undefined') { -%>
"hostname": "<%=hostname%>",
<% } -%>
```

**Comment by WangWinson** [comment]
:+1: after comments addressed

File: data/templates/photon-os-ks (hunk 8 above)
Comment (cgx027) [inline data/templates/photon-os-ks]: Moved the ```if``` statement above ```postInstall``` and it looks better now.

**Comment by panpan0000** [comment]
test this please

**Comment by iceiilin** [comment]
:+1:

**Comment by yyscamper** [comment]
jenkins: test his please

**Comment by cgx027** [comment]
test this please

**Comment by yyscamper** [comment]
test this please

**Comment by yyscamper** [comment]
Forgot my :+1:

**Pull request merged**